text/0000-unicode-idents.md

```diff
@@ -0,0 +1,145 @@
+- Feature Name: unicode_idents
+- Start Date: 2018-06-03
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow non-ASCII letters (such as accented characters, Cyrillic, Greek, Kanji, etc.) in Rust identifiers.
+
+# Motivation
+[motivation]: #motivation
+
+Rust is written by many people who are not fluent in the English language. Using identifiers in ones native language eases writing and reading code for these developers.
```

Comment:
To have proper support then they should be allowed, as it affects the rendering of text, which may cause it to have different meanings (I'm not sure, but it seems likely to me). However, this can cause issues like in C++ and Swift for example， different number of zero-width joiners or spaces are different identifiers, and that leads to a readability nightmare.

Reply:
That is why UAX <span>#</span>31 suggests allowing them only in certain positions, to avoid having multiple distinct identifiers look identical.

Reply:
So AIUI these are used for:


  - Forcing explicit viramas in Indic scripts in consonant clusters. There _is_ a semantic difference but it's exceedingly minor and mostly comes in play for Sanskrit. It's a pretty minor difference; AIUI it's used in certain kinds of compound words and is very easily omitted

 - Certain vowel presentational forms in Bengali and Oriya. Also minor.

 - Forcing letters to take different (word-medial, etc) forms in the Perso-Arabic script, used for:
   - Arabic affixes, when shown in isolation or when used with non-arabic words. Imagine you had to write something like "Rust's" where "Rust" is in the Latin script but the 's is in Arabic, you need the Arabic suffix to not use a word-initial text form. An example of this is the [ب  prefix preposition](https://en.wiktionary.org/wiki/%D8%A8#Preposition)

   - [Arabic abbreviations](https://github.com/w3c/alreq/issues/91)

AFAICT these are all somewhat optional (though preferred). The abbreviations one might be the most used.


(that said, there are much larger problems with using an RTL script in rust)

Reply:
ZWNJ has important use in Persian at least https://en.m.wikipedia.org/wiki/Zero-width_non-joiner